File: totaleconomy-common/src/main/java/com/ericgrandt/totaleconomy/common/command/BalanceCommand.java
package com.ericgrandt.totaleconomy.common.command;

import com.ericgrandt.totaleconomy.common.dto.CurrencyDto;
import com.ericgrandt.totaleconomy.common.econ.CommonEconomy;
import com.ericgrandt.totaleconomy.common.game.CommonPlayer;
import com.ericgrandt.totaleconomy.common.game.CommonSender;
import java.math.BigDecimal;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import net.kyori.adventure.text.Component;

public class BalanceCommand implements CommonCommand {
    private final CommonEconomy economy;
    private final CurrencyDto currency;

    public BalanceCommand(final CommonEconomy economy, final CurrencyDto currency) {
        this.economy = economy;
        this.currency = currency;
    }

    @Override
    public boolean execute(CommonSender sender, Map<String, CommonParameter<?>> args) {
        if (!(sender instanceof CommonPlayer player)) {
            return false;
        }

        CompletableFuture.runAsync(() -> onCommandHandler(player));

        return true;
    }

    private void onCommandHandler(CommonPlayer player) {
        BigDecimal balance = economy.getBalance(
            player.getUniqueId(),
            currency.id()
        );
        if (balance == null) {
            player.sendMessage(Component.text("No balance found"));
            return;
        }

        player.sendMessage(Component.text("Balance: ").append(
            economy.format(currency, balance)
        ));
    }
}

File: totaleconomy-common/src/main/kotlin/com/ericgrandt/totaleconomy/command/BalanceCommand.kt
package com.ericgrandt.totaleconomy.command

import com.ericgrandt.totaleconomy.econ.CommonEconomy
import com.ericgrandt.totaleconomy.game.CommonPlayer
import com.ericgrandt.totaleconomy.game.CommonSender
import com.ericgrandt.totaleconomy.model.errorToUserMessage
import com.github.michaelbull.result.mapBoth
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import net.kyori.adventure.text.Component

class BalanceCommand : CommonCommand {
    private val economy: CommonEconomy

    constructor(economy: CommonEconomy) {
        this.economy = economy
    }

    override fun runAsync(
        scope: CoroutineScope,
        sender: CommonSender,
        args: MutableMap<String, CommonParameter<*>>
    ): Boolean {
        if (sender !is CommonPlayer) {
            return false
        }

        scope.launch {
            execute(sender)
        }

        return true
    }

    private fun execute(player: CommonPlayer) {
        economy.getBalance(player.getUniqueID()).mapBoth(
            success = {
                player.sendMessage(Component.text("You have ").append(economy.format(it)))
            },
            failure = {
                val userMessage = errorToUserMessage(it)
                player.sendMessage(Component.text(userMessage))
            }
        )
    }
}